### ISSUE TITLE
Sensor calibration dependent on logger

### ISSUE THREAD

**Issue by abohara** [IEA-Task-43/digital_wra_data_standard#205]
Some sensors e.g. NRG T60 (temperature) or 200M (wind vane) report different slope / offset to be used dependent on the type of loggers (e.g. SymphoniePlus or SymphoniePro ). I don't believe our calibration data model supports this discrimination. Or how this should be reported in situations where one logger type has been swapped for another ( which is not that uncommon) and the sensors remain the same.

Any thoughts ? @stephenholleran @kersting @heikowestermann

**Comment by kersting** [comment]
@abohara wouldn't this be covered by the `slope` and `offset` of `logger_measurement_config`? If someone replaces the logger new constants will be entered in that logger.

**Comment by stephenholleran** [comment]
I think what @abohara is asking is how do we distinguish in the sensor.calibration area. Correct me if I'm wrong?

I can see this been a problem alright but not sure how common it would be?

To solve this we could have the 'logger_model_name' field also in the calibration. Then the user can report all the slope and offsets for each logger. During an analysis, the information is there for the user, or code, to assess if an adjustment is needed.

**Comment by heikowestermann** [comment]
@abohara @stephenholleran @kersting 
We do not distinguish between different loggers for our calibrations because we do not know what type of logger the user will use in the future. The 200M has a "default slope and offset" according to the [data sheet](https://www.nrgsystems.com/assets/resources/NRG-200m-Tech-Product-Sheet-2020-V1.pdf) which matches the output of 0-2.5V for 0-360 deg and our calibrations. I know the SymphonyPRO loggers have a different default slope/offset, but I don't know the details or conversion factors for that. We don't currently use them. Perhaps @sdsmdp has more details?

I would suggest that the slope/offset/table for a given instrument should be converted (if necessary) at the logger level and not at the calibration level, since this is logger dependent and not instrument dependent, and the information about which logger is being used may not be available when the calibration reports are generated.

**Comment by abohara** [comment]
Thanks @heikowestermann for your comments. @kersting  and @stephenholleran agree with the proposed solution

**Comment by stephenholleran** [comment]
Hi All,

@abohara I'm not sure @heikowestermann actually proposed a solution. Just that it should probably go in the logger area as it is logger specific, which is a fair point. I'm not sure how that would look though.

I think, but needs confirmation, that all loggers are either 2.5V or 5V. There could be others I'm sure. What I am getting at is that the Voltage is really the important thing. So instead of noting the logger in the DCC which @heikowestermann won't know, could the Voltage or Voltage range that was used for the calibration be noted? That way a conversion could be performed or the mast installer will know which slope/offset they need to type into the particular logger they are using.

@abohara does the calibration doc for the 200M or T60 also report the voltage range for each slope/offset?

**Comment by sdsmdp** [comment]
@abohara, I spoke with my chief engineer regarding this topic today. He confirmed what I was trying to explain on the call y'day. The good news is that he provided me with some better language to convey it. I will write that up ASAP and post back.

**Comment by abohara** [comment]
Thanks @sdsmdp . Looking forward to it.

**Comment by abohara** [comment]
Hi @sdsmdp . Any updates on this ? Thank you!

**Comment by kersting** [comment]
@abohara, @sdsmdp, @stephenholleran, and @heikowestermann should we schedule a meeting to tackle this? It seems that this a recurrent thing that we see specially for the NRG systems sensor/logger combo. Stephen's suggestion probably could work but I think we need to discuss some of the pitfalls and what is required. I can send a doodle poll if you agree with the idea of the meeting.

**Comment by stephenholleran** [comment]
We can take the old slot of our usual calls this Thursday. 5 pm Irish time? Or if that is too close, the same slot in 2 weeks time at either 4 or 5 pm Irish time. You could send a doodle poll for those?

**Comment by sdsmdp** [comment]
I responded to the poll. I followed up with my engineers at the end of January but have not had time to write and share my notes with you. The meeting is an excellent forcing function.

Although I can take time to explain why the slope and offset differ depending on the logger model, the main point is that this is a legitimate use case. Please let me know if you'd prefer the meeting time to understand why this is the case (I should prepare to present) or to focus on a solution for handling this use case in the model. Thank you.

**Comment by abohara** [comment]
Thanks @sdsmdp . Agree with @kersting . An explanation would be helpful and just discussion on this topic would be sufficient ( no detailed presentation needed ).

**Comment by stephenholleran** [comment]
## SH notes from call between @kersting, @sdsmdp, @heikowestermann and myself.

Below is a screenshot of the specs for the W200M wind vane from NRG Systems. This shows that there are two transfer functions depending on two different logger types. Slope 147.91°/V, offset 1.460° for SymPRO and slope 0.368°/V, offset -5.3° for SymPLUS3.
https://www.nrgsystems.com/products/met-sensors/detail/nrg-200m-wind-direction-vane/
![image](https://user-images.githubusercontent.com/25622575/227278713-d2118535-c5c5-4150-bfc7-68075190df62.png)

### My understanding:

The reason for these different slope and offsets depending on the logger is because part of the transfer from an analog signal to degrees is hidden by the logger. There are 2 steps that the logger goes through as outlined below.

1. First the logger needs to convert the analog voltage values into a digital signal via a converter, aka [A-to-D](https://en.wikipedia.org/wiki/Analog-to-digital_converter). This A-to-D can be different for all logger types and primarily relies on the precision or resolution that the logger can operate which also means how many bits can it handle. For example a 10 bit logger can store a value up to 1,024 which means for this wind vane that outputs an analog voltage in the range of 0 to 2.5 V the 10 bit logger converts this to a digital range to 0 to 1,024. Therefore if the wind vane outputs 2.5 V the A-to-D converter outputs 1,024. If the wind vane outputs 2 V, this results in 1,024/2.5 x 2 = 819.
2. Secondly this digital value is converted to degrees by apply a slope and offset. The 1,024 value (resulting from an output of 2.5 V) is converted to 360 degrees using the appropriate slope and offset.

Another point to note is that when a calibration facility provides a slope and offset resulting from a calibration, this is the slope and offset that converts the analog voltage directly to degrees.

Mentioned above are two different slope and offsets, one in step 2 of the logger conversion and the second provided by the calibration converting volts to degrees.
- The SymPro logger requires the slope and offset coming directly from the calibration which converts volts directly to degrees. For this wind vane that is a slope of 147.91°/V and offset 1.460°. A wind vane output of 2.5 V results in 2.5 x 147.91 + 1.46 = 371 degrees (_note I don't know why it doesn't give exactly 360 degrees but for this explanation it is close enough. There could be provisions for cable length or other specifics to electronics I am not aware of._) For an output of 2 V results in 2 x 147.91 + 1.46 = 297 degrees
- Whereas the SymPLUS3 logger requires the slope and offset from step 2 in the conversion. For this wind vane that is a slope of 0.368°/V and offset -5.3°. The full conversion equation is 1024/2.5 x 2.5 x 0.368 - 5.3 = 371 degrees. For an output of 2 V results in 1024/2.5 x 2 x 0.368 - 5.3 = 296 degrees

I do not know the reason why for the SymPRO the first step in the logger conversion is "hidden" but for a wind analyst, thinking only of volts to degrees, this is intuitive and it also matches the calibration certificates.

@sdsmdp please can you confirm or correct my understanding? Cheers.

cc @abohara

**Comment by sdsmdp** [comment]
That was pretty close, @stephenholleran . It may be fairer to say that I think you accurately captured what I was trying to say verbally. I don't think I did a great job explaining, though. 

Let me try to write it the general principle in a way that hopefully makes sense:

Analog sensors generate continuous signals that represent physical quantities such as temperature, pressure, or light intensity. These signals are usually in the form of voltage or current. The analog signals must be converted to digital format using a data logger to record and process the sensor data. Different data loggers have different specifications and capabilities and thus require different scaling for the analog sensor signals.

One key factor determining the scaling of an analog sensor signal is the bitness of the data logger's Analog-to-Digital Converter (ADC). The ADC is responsible for converting the continuous analog signal into a discrete digital value that can be processed and stored.

The bitness of an ADC refers to the number of bits used to represent the digital value of the input analog signal. Common ADC bitness values are 8, 12, 16, or 24 bits. The bitness directly impacts the resolution of the digital signal, with higher bitness providing a finer resolution. The resolution can be defined as the smallest change in the input signal that the ADC can detect.

For example, a 12-bit ADC has a resolution of 2^12, or 4096 discrete levels, while a 16-bit ADC has a resolution of 2^16 or 65536 discrete levels. This means that a 16-bit ADC can detect smaller changes in the input signal compared to a 12-bit ADC.

To properly scale an analog sensor signal for a specific data logger, one needs to consider the ADC bitness and the input voltage range of the data logger. The goal is to map the entire range of the sensor output to the full range of the ADC, ensuring that the digital values are proportional to the analog input signal.

For example, let's say we have a temperature sensor with an output range of 0-5V and a data logger with a 12-bit ADC and a 0-5V input range. We can directly connect the sensor to the data logger, and the ADC will convert the temperature signal with a resolution of 5V/4096, or approximately 1.22mV per digital step.

However, the scaling needs to be adjusted if we connect the same sensor to a data logger with a 16-bit ADC and a 0-10V input range. Without scaling, the ADC would use only half of its range (0-5V out of 0-10V), resulting in a reduced resolution. To maximize the resolution, an amplifier could be used to scale the sensor output to match the 0-10V range of the data logger. The amplified signal would then utilize the full range of the 16-bit ADC, providing a higher resolution of 10V/65536, or approximately 0.153mV per digital step.

SymphoniePRO employs a 16-bit ADC, offering 76µV resolution over a 5V input range. Accuracy is 0.04% of full scale. As you can see from NRG analog sensor specifications and calibration reports, this results in a scaling requirement that is on par with what is measured at calibration time. SymphoniePLUS3 uses a 10-bit ADC, offering 2.44mV resolution over a 2.5V input range. Accuracy is 0.1% of full scale. This requires greater scaling than SymPRO to provide the same physical measurements.

Additional factors, such as long sensor cable lengths, can significantly impact the accuracy and reliability of analog sensor measurements. There are several factors to consider when using long cables to connect analog sensors to a data logger:

1. Signal attenuation: As the cable length increases, the analog signal may experience attenuation due to the cable's inherent resistance, inductance, and capacitance. This can result in a weaker signal reaching the data logger, which can lead to reduced measurement accuracy and resolution. One can use low-resistance cables or opt for shorter cable runs when possible to mitigate this issue.
 
2. Noise pickup: Long cables can act as antennas, picking up electromagnetic interference (EMI) from nearby sources such as power lines, motors, or other electronic devices. This noise can be coupled to the sensor signal, causing inaccurate or erratic measurements. Shielded cables can be used to minimize noise pickup, and properly grounding the shield at the data logger end is crucial. In addition, twisted-pair cables can help to reduce the impact of EMI by canceling out the common-mode noise picked up along the cable length.

3. Voltage drop: A long cable introduces resistance, which can cause a voltage drop along the cable length when the current is flowing. This voltage drop can lead to inaccuracies in measurements, especially for sensors that output current signals (e.g., 4-20 mA sensors). To counteract this issue, one can use thicker cables with lower resistance or opt for voltage output sensors that are less affected by voltage drops.
 
4. Signal reflection: When using high-frequency or fast-changing signals, long cable lengths can cause signal reflections due to impedance mismatches between the cable, sensor, and data logger. These reflections can distort the signal waveform, leading to inaccurate measurements. To minimize signal reflections, impedance matching techniques can be applied, or slower-changing signals can be used if possible.
 
5. Temperature effects: Long cables can experience temperature variations along their length, which can affect the cable's resistance and the accuracy of the sensor measurement. This is particularly important for temperature-sensitive sensors like thermocouples or resistance temperature detectors (RTDs). To minimize temperature-related inaccuracies, selecting cable types with low-temperature coefficients or using compensation techniques to account for temperature-related errors is important.

NRG, for example, indicates additional scaling associated with the cables we sell. It has to be specific based on the properties of the cabling we select and how we prepare it. We most often reflect this additional scaling as an additional offset per meter of cabling.

In summary, analog sensors need different scaling depending on the data logger they are connected to because of variations in ADC bitness and input voltage ranges. Other factors, such as cable length, also affect scaling. Proper scaling ensures that the sensor output is matched to the data logger's capabilities, maximizing the resolution and accuracy of the acquired data.

This is not altogether different cc

**Comment by stephenholleran** [comment]
## Notes from call with NRG Systems engineers

1. There is a simple algorithm to convert the calibration slope & offset provided by a calibration facility to the one that needs to be typed into the SymPLUS3 logger.
2. Instead of trying to store multiple calibration slope & offsets depending on the logger the sensor "might" be installed on, an open-source piece of code could be shared which can make the adjustment from calibration slope & offset to what needs to be typed into the logger.
3. A wind analyst then compares those numbers and not the calibration slope & offset directly.
4. AB and GK to send a list of sensors to NRG Systems so they can narrow the scope of that algorithm development.

Thanks to the NRG Systems engineers that came on the call!

cc @abohara @kersting @heikowestermann @sdsmdp @MatteoManaRWE

**Comment by abohara** [comment]
Thanks for the notes @stephenholleran . I have it on my plate to get a list of sensors over

**Comment by stephenholleran** [comment]
Hi All,

Just coming back to this.

@kersting I believe you sent a list of sensors to NRG and I think there was a question about just NRG sensors or all sensors.

Any developments from NRG Systems on the development of that algorithm to convert the calibration slope & offset to the one the needs to be typed into the SymPLUS3 logger?

cc @sdsmdp

**Comment by kersting** [comment]
@stephenholleran the only updates that I got was that NRG was not going to provide any conversion for other sensors. @abohara  sent a list of NRG sensors to NRG. @abohara any news on that?

**Comment by kersting** [comment]
> @stephenholleran the only updates that I got was that NRG was not going to provide any conversion for other sensors. @abohara sent a list of NRG sensors to NRG. @abohara any news on that?

Based on the comments above, I recommend closing this issue.

**Comment by abohara** [comment]
@kersting - Nothing from my end. I support closing it as well!
